File: safehouses/7916243/safehouse.py

def dist(p1, p2):
    return abs(p2[0] - p1[0]) + abs(p2[1] - p1[1])

def main():
    N = int(input())
    safehouses = []
    spies = []

    for r in range(N):
        line = input()
        for c in range(N):
            if line[c] == 'H':
                safehouses.append((r,c))
            elif line[c] == 'S':
                spies.append((r,c))

    max_dist = 0
    for s in spies:
        nearest_dist = float('inf')
        for h in safehouses:
            d = dist(s, h)
            if d < nearest_dist:
                nearest_dist = d
        max_dist = max(nearest_dist, max_dist)

    print(max_dist)

if __name__ == '__main__':
    main()

File: safehouses/7918395/Safehouse.java
import java.util.Scanner;
import java.util.ArrayList;

public class Safehouse {
    public static class Point {
        public int x;
        public int y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }

    public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);

        int N = scan.nextInt();
        ArrayList<Point> safehouses = new ArrayList<>();
        ArrayList<Point> spies = new ArrayList<>();

        for (int r = 0; r < N; ++r) {
            String line = scan.next();
            for (int c = 0; c < N; ++c) {
                if (line.charAt(c) == 'H') {
                    safehouses.add(new Point(r, c));
                } else if (line.charAt(c) == 'S') {
                    spies.add(new Point(r, c));
                }
            }
        }

        scan.close();

        int maxDist = spies.stream().map((s) -> safehouses.stream().map((h) -> dist(s, h)).min(Integer::compare).get()).max(Integer::compare).get();
        System.out.println(maxDist);
    }

    public static int dist(Point a, Point b) {
        return Math.abs(a.x - b.x) + Math.abs(a.y - b.y);
    }
}
